src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+}
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size) {
+  LOG_TRACE("Enter");
+  size_t h = reinterpret_cast<HashCodeType>(hashcode);
+  return h % size;
+}
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size) {
+  LOG_TRACE("Enter");
+  return (home + offset) % size;
+}
+
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
+                   const KeyType key,
+                   const HashCodeType hashcode,
+                   const size_t home) {
+  LOG_TRACE("Enter");
+  size_t i = 0;
+  size_t size = tmp_buckets.size();
+  for (i = 0; i < size; ++i) {
+    size_t real_index = get_probe(home, i, tmp_buckets.size());
+    SequentialBucket &bkt = tmp_buckets[real_index];
+    // If not found
+    if (bkt.is_empty()) {
+      // This is first, because equality on an empty bucket is not well defined.
+      return {FOUND_HOLE, i};
+    } else if (bkt.offset < i) { // This means that bkt belongs to a nearer home
+      return {FOUND_SWAP, i};
+    // If found
+    } else if (bkt.equal_key(key, hashcode)) {
+      return {FOUND_MATCH, i};
+    }
+  }
+  return {FOUND_NOHOLE, SIZE_MAX};
+}
+
+int
+insert_without_resize(      std::vector<SequentialBucket> &tmp_buckets,
+                          const KeyType key,
+                          const ValueType value,
+                          const HashCodeType hashcode) {
+  LOG_TRACE("Enter");
+  SequentialBucket tmp = SequentialBucket(key, value, hashcode, /*random value*/0);
+  // This could also be upper-bounded by the number of valid elements (num_elem)
+  // in tmp_buckets. This is because you need to bump at most num_elem elements
+  // (if they are all sitting in a row) to insert something.
+  while (true) {
+    size_t home = get_home(tmp.hashcode, tmp_buckets.size());
+    const auto [status, offset] = get_wouldbe_offset(tmp_buckets, tmp.key, tmp.hashcode, home);
+    switch (status) {
+      case FOUND_MATCH: {
+        LOG_DEBUG("FOUND_MATCH");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        bkt.value = tmp.value;
+        return 0;
+      }
+      case FOUND_SWAP: {
+        LOG_DEBUG("FOUND_SWAP");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        tmp.offset = offset;
+        bkt.swap(tmp);
+        // TODO(dchu): likely to be buggy
+        continue;
+      }
+      case FOUND_HOLE: {
+        LOG_DEBUG("FOUND_HOLE");
+        size_t real_index = get_probe(home, offset, tmp_buckets.size());
+        SequentialBucket &bkt = tmp_buckets[real_index];
+        tmp.offset = offset;
+        bkt.swap(tmp);
+        // TODO(dchu): likely to be buggy
+        return 0;
+      }
+      case FOUND_NOHOLE:
+        assert(0 && "no hole found!");
+        exit(1);
```

Comment:
What would you recommend? I want this to be a fatal error and exit, sort of like Rust's `panic` (if my very cursory understanding is correct)

Reply:
Honestly, I'm not sure. Does reaching this condition imply an internal bug? Could it also be cause by an OOM error?

Either way, my opinion is that we should prioritise allowing destructors to be called so that whatever resources users may have will be released. We have no idea what resources they might have, so I'd rather assume it can't just be handled by the OS. In which case, I'm leaning towards throwing an exception, which actually isn't too dissimilar to `panic`. Unfortunately, the line between recoverable exceptions and unrecoverable exceptions isn't very clear in C++ so people might catch it and try to do something with it, which wouldn't be great.

The other option is: if we have determined this is only ever an internal bug/developer error, then we could stick `__builtin_unreachable()` (or Windows `__assume(false)`) there to satisfy the compiler.

Reply:
I agree with the statement exceptions aren't enough.

I think it goes beyond just `__builtin_unreachable()`. If we do reach there (some bug), we'd want to panic. I guess I'll just leave the assertions and take out the exit calls. In the past, I've just defined a macro `IMPOSSIBLE()` and then we can decide what to do there later. But idk if everyone wants to use **David's Ingenious Macros With Impossibility Traps**...

P.S. What would we be doing that the OS can't clean up?

Reply:
> **D**avid's **I**ngenious **M**acros **W**ith **I**mpossibility **T**raps

Hmm that's an interesting name for those macros...

But yeah, I'm probably bikeshedding here. The macro actually isn't a bad idea though since it does help express intent.

> P.S. What would we be doing that the OS can't clean up?

Not sure. I'm just thinking in terms of wider use cases, not necessarily constrained to only what we're doing. Who knows, maybe users might be in the middle of a transaction with an external server. Ideally they should be able to define what happens on error - whether a rollback or something else.

Reply:
I love the commitment to portability. How about we switch to C89?

Reply:
Bruhh C89?? Don't make me have to declare all my variables at the beginning of the function 😱